hw/ip/flash_ctrl/data/flash_ctrl.sv.tpl

```diff
@@ -529,8 +529,12 @@ module flash_ctrl
 
   // a read request is seen from software but a read operation is not enabled
   // AND there are no pending entries to read from the fifo.
+  // This indicates software has issued a read when it should not have.
+  //
+  // sw_sel qualification is used here to ensure the no_op condition is ignored
+  // when software is not selected through arbitration.
   logic rd_no_op_d, rd_no_op_q;
-  assign rd_no_op_d = rd_fifo_ren & ~rd_op_valid & ~sw_rvalid;
+  assign rd_no_op_d = rd_fifo_ren & ~rd_op_valid & ~sw_rvalid & sw_sel;
```

Comment:
Please help me to understand. I am seeing `rd_no_op_d` affects tlul adapter grant and `rvalid`. So, intention is to complete the TL-Ul transaction when there's no operation at all. `sw_rvalid` and `adapter_rvalid` are the actual path when there's op.
Do I understand correctly?

Reply:
The intention is to send a bus error back to the requester when the request would never complete (avoiding a hung bus). `rd_no_op_d` identifies that case (request is a read, there's no ongoing read operation, and the FIFO is empty).

This PR adds a bit to avoid/delay triggering the error if the `flash_ctrl_arb` is busy doing work for the "hardware interface", since `rd_op_valid` won't be high in that case, even if an operation from the "software interface" is actually pending on the arbiter.

And yup, `sw_rvalid` and `adapter_rvalid` are the normal path.

Reply:
thanks for explaining @a-will